src/lib.rs

```diff
@@ -56,7 +56,13 @@
 //!
 //! let btf = Btf::from_file("/sys/kernel/btf/vmlinux").unwrap();
 //!
-//! let func = match btf.resolve_types_by_name("kfree_skb_reason").unwrap().pop().unwrap() {
+//! let func = match btf
+//!     .resolve_types_by_name("kfree_skb_reason")
+//!     .unwrap() // Result
+//!     .unwrap() // Option
```

Comment:
I'm wondering if we could make it a bit easier to use by using vector with 0 length as an empty result. This would avoid at least 1 `unwrap`

Reply:
Or, as we're breaking the API anyways, how about returning an iterator over the results? It also has a clear semantics for zero result (`next()` returns `None`) and we could be more performant (avoid looking in all the BTF file if we just ant the first occurrence). All of this, of course, could be done in a follow up PR.

Reply:
> I'm wondering if we could make it a bit easier to use by using vector with 0 length as an empty result. This would avoid at least 1 `unwrap`. WDYT?

In normal use cases the above would look better, using `?`. But yes, I'm wondering the same too :) I don't know, I don't have a clear preference for now but it's definitively something to thing about (before merging this).

> Or, as we're breaking the API anyways, how about returning an iterator over the results? It also has a clear semantics for zero result (`next()` returns `None`) and we could be more performant (avoid looking in all the BTF file if we just ant the first occurrence). All of this, of course, could be done in a follow up PR.

Right; please see my previous answer about that:

> That sounds good but then I'm not sure how this would apply. We do have an iterator over linked types because they are like a linked list and we can do one lookup at a time. But for all the other resolving helpers it always boils down to resolving type ids, which are already stored as a vec in `BtfObj`.
> 
> So we would either:
> 
>     * Hide a Vec into an iterator, which is not helping, except for maybe providing a better API?
> 
>     * Rework helpers in `BtfObj` to look for the next type id in the internal vector. I'm not sure of the benefits here, it would be slower and as we already store those id in memory, by design the list has to be limited in size already.
> 
> 
> BTW, I do like the iterator from the `btf-rs` user's point of view, it would be quite clean. But I'm really not sure it provides value at other levels.
> 
> Thoughts?

Reply:
Sorry, I must have missed that comment. I agree it's probably an overkill to rework all the internals just to provide an interator API.

I slightly lean towards using the vector of zero size to indicate zero elements found, basically because if I put myself on the user's shoes, I would ask: OK, if `Error` means error, and `Ok(None)` means no elements, what does `Ok(Some(zero-size-vector))` mean?

Reply:
> Sorry, I must have missed that comment. I agree it's probably an overkill to rework all the internals just to provide an interator API.

I'm not against doing so tbf (in another effort), but my main concern is how to implement this so it's actually an iterator and not a "fake" one (aka just iterating over a list we already have).

> I slightly lean towards using the vector of zero size to indicate zero elements found, basically because if I put myself on the user's shoes, I would ask: OK, if `Error` means error, and `Ok(None)` means no elements, what does `Ok(Some(zero-size-vector))` mean?

Yeah, that's what puzzles me as well (although that can't happen, but Rust does not know that) :) It's probably best to go with not using an `Option`.

Reply:
I went with not using an `Option` when a `Vec` is returned.